tripal/t4d8
# Upgrade blocks

# New Feature

Issue #18 

## Description

 - Upgraded implementation of Powered By Tripal, Notifications, and Content Type Bar Chart blocks.
 - Added Tripal Logo images to `tripal/images` folder.
 - The `admin/dashboard` page no longer exists by default in Drupal, so I added it to `tripal/tripal.routing.yml`
 - Added `tripal/css/tripal.css`, which is nearly the same as Tripal 3's [`tripal/theme/css/tripal.css`](https://github.com/tripal/tripal/blob/7.x-3.x/tripal/theme/css/tripal.css)
 - Separated D3 JS library from Tripal JS library and updated local D3 JS (`tripal/js/d3.min.js` -> `tripal/js/d3.v3.min.js`).
 - Added Tripal Dashboard JS library to `tripal/tripal.libraries.yml`.
 - Added `tripal_bundle` table.


## Testing

 - Re-install Tripal 4 in order to create the `tripal_bundle` table.
 - Clear your cache
 - Add some data to the `tripal_bundle` table and a `chado_bio_data_N` table, for example:
```SQL
INSERT INTO tripal_bundle (type, term_id, name, label) VALUES
    ('TripalEntity', 2, 'bio_data_2', 'Analysis'),
    ('TripalEntity', 3, 'bio_data_3', 'Organism');
CREATE SEQUENCE chado_bio_data_2_mapping_id_seq START 1;
CREATE SEQUENCE chado_bio_data_3_mapping_id_seq START 1;
CREATE TABLE chado_bio_data_2 (
    mapping_id INTEGER NOT NULL DEFAULT nextval('chado_bio_data_2_mapping_id_seq'::regclass) 
    PRIMARY KEY,
    entity_id INTEGER NOT NULL,
    record_id INTEGER NOT NULL,
    nid INTEGER);
CREATE TABLE chado_bio_data_3 (
    mapping_id INTEGER NOT NULL DEFAULT nextval('chado_bio_data_3_mapping_id_seq'::regclass) 
    PRIMARY KEY,
    entity_id INTEGER NOT NULL,
    record_id INTEGER NOT NULL,
    nid INTEGER);
INSERT INTO chado_bio_data_2 (mapping_id, entity_id, record_id) VALUES
    (1, 38496, 7),
    (2, 26551, 5),
    (3, 4786, 2);
GRANT ALL PRIVILEGES ON ALL TABLES IN SCHEMA public TO drupal;
```
 - Add some data to the `tripal_admin_notfications` table, for example:
```SQL
INSERT INTO tripal_admin_notfications (details, title, actions, submitter_id, type) VALUES
('some details', 'title', 'a:1:{s:15:"Go to dashboard";s:16:"tripal.dashboard";}', 1, 'Field');
```
 - Navigate to `admin/dashboard`
 - Verify that the notifications and content type bar chart blocks both work as expected.
 - Navigate to `admin/structure/block`
 - Choose a block layout area and click "Place block"
 - In the menu, scroll down to the "Powered by Tripal" block and click "Place block"
 - Choose your logo size and type, and your visibility restrictions, then click "Save block"
 - Navigate to a page which satisfies your visibility restrictions you set in the previous step, and verify that the Powered by Tripal block appears on the page.

## Additional Notes

I am waiting to mark this PR as "ready to merge" because I would like to have the `tripal_admin_notfications` table spelled correctly (PR #42) before this code is merged, since the Notifications block depends on that table.

Also, the "Dismiss notification" links in the Notifications block currently do nothing, so I'm wondering if those links count as part of the block upgrade issue (#18). If so, I can implement that and add it to this PR.
